{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "source": [
    "import boto3\r\n",
    "import pandas as pd\r\n",
    "import numpy as np\r\n",
    "import re"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "source": [
    "client = boto3.client('s3')\r\n",
    "\r\n",
    "# sequence file\r\n",
    "\r\n",
    "path = 's3://transcriptional-effect-adversity-in-hippocampus/sra_metadata_07dae975-66fa-4503-a5ad-bdbc99bb2fef.csv'\r\n",
    "\r\n",
    "path_sample = 's3://transcriptional-effect-adversity-in-hippocampus/sample_1.fastq'\r\n",
    "\r\n",
    "path_13062383 = 's3://transcriptional-effect-adversity-in-hippocampus/SRR13062383_1.fastq'"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "source": [
    "df = pd.read_csv(path_sample, header=None, sep='\\n')\r\n",
    "# df = pd.read_csv(path_sample, error_bad_lines=False) # This skips the line with error when loading data\r\n",
    "pd.set_option('display.max_colwidth', None)\r\n",
    "df.head(10)"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "                                                                                                                                                        0\n",
       "0                                                                                                                             @SRR13062383.1 1 length=150\n",
       "1  CCCTGCCATCCGGAACCACGACAAGGTCCTGCAGCGCCTGTGTGAGCACTTGGACCACGCCCTGTTGTACGGATTGCAAGACCTGTCCTCTGGCTACTGGGTGGTCGTGGTCCATTTCACCCGCAGAGAGGCCATCAGGCAGATCGGAAG\n",
       "2                                                                                                                             +SRR13062383.1 1 length=150\n",
       "3  BBBDFFFFDHHFDCGHGIJGIIJJFDFHHIIJJHHGIJHIGIHIJJJIJGGHHHHFEFDDDDBDDDDCCDDBDDDDDDDCDDDDDDDDDDDDDDDDDDCDD95>5CD@DB?CCCEEEEDDDDDDDDDDCDBDDDDDDCCDDDDDDDDDDB\n",
       "4                                                                                                                             @SRR13062383.2 2 length=150\n",
       "5  NTCGACTCCATTAAAGACTTTTCTTTAAAATAGCAGAGCCTGAACCTTGATATCCAAGCCTTAGCCTGCCCGTCCGCCCTGCCTATGATCATAGAGCTCCATCGGTTCAAAGAACCCCAGACATCGGCAGCCCACCCCTCTGAACTCCAG\n",
       "6                                                                                                                             +SRR13062383.2 2 length=150\n",
       "7  #1=+<+@)2@2,,+<+2<<C>7,33@+?7+2+*:*@:?3?7))**:3A*0*0?BB###############################################################################################\n",
       "8                                                                                                                             @SRR13062383.3 3 length=150\n",
       "9  NTCCTTCTCTCTCTGGTTTAACTTATCTTGTACTAGAAAAGCCAGGGAGGTGGTAAATGTATGGTGGTGTTTTGCTTACAATAGTTCACAAAAATCTCTCTCTAGGATGAGATGCTGGCACTGACGAACATTGCCTAAGATTTCTATGAA"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>@SRR13062383.1 1 length=150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCCTGCCATCCGGAACCACGACAAGGTCCTGCAGCGCCTGTGTGAGCACTTGGACCACGCCCTGTTGTACGGATTGCAAGACCTGTCCTCTGGCTACTGGGTGGTCGTGGTCCATTTCACCCGCAGAGAGGCCATCAGGCAGATCGGAAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>+SRR13062383.1 1 length=150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BBBDFFFFDHHFDCGHGIJGIIJJFDFHHIIJJHHGIJHIGIHIJJJIJGGHHHHFEFDDDDBDDDDCCDDBDDDDDDDCDDDDDDDDDDDDDDDDDDCDD95&gt;5CD@DB?CCCEEEEDDDDDDDDDDCDBDDDDDDCCDDDDDDDDDDB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>@SRR13062383.2 2 length=150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NTCGACTCCATTAAAGACTTTTCTTTAAAATAGCAGAGCCTGAACCTTGATATCCAAGCCTTAGCCTGCCCGTCCGCCCTGCCTATGATCATAGAGCTCCATCGGTTCAAAGAACCCCAGACATCGGCAGCCCACCCCTCTGAACTCCAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>+SRR13062383.2 2 length=150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>#1=+&lt;+@)2@2,,+&lt;+2&lt;&lt;C&gt;7,33@+?7+2+*:*@:?3?7))**:3A*0*0?BB###############################################################################################</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>@SRR13062383.3 3 length=150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NTCCTTCTCTCTCTGGTTTAACTTATCTTGTACTAGAAAAGCCAGGGAGGTGGTAAATGTATGGTGGTGTTTTGCTTACAATAGTTCACAAAAATCTCTCTCTAGGATGAGATGCTGGCACTGACGAACATTGCCTAAGATTTCTATGAA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "metadata": {},
     "execution_count": 4
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "source": [
    "# bed file\r\n",
    "\r\n",
    "path_local_bed1 = \"./Chip-seq/GSM2228982_42YVE_s_5_sequence_norm.bed/GSM2228982_42YVE_s_5_sequence_norm.bed\"\r\n",
    "\r\n",
    "df_bed_chr1 = pd.read_csv(path_local_bed1, header=None, sep='\\n')\r\n",
    "# df = pd.read_csv(path_sample, error_bad_lines=False) # This skips the line with error when loading data\r\n",
    "pd.set_option('display.max_colwidth', None)\r\n",
    "df_bed_chr1.head(10)"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "                               0\n",
       "0    chr1\\t67926\\t67962\\t0\\t1\\t+\n",
       "1    chr1\\t97370\\t97406\\t0\\t1\\t+\n",
       "2  chr1\\t101069\\t101105\\t0\\t1\\t-\n",
       "3  chr1\\t132314\\t132350\\t0\\t1\\t-\n",
       "4  chr1\\t156509\\t156545\\t0\\t1\\t-\n",
       "5  chr1\\t157123\\t157159\\t0\\t1\\t+\n",
       "6  chr1\\t165094\\t165130\\t0\\t1\\t+\n",
       "7  chr1\\t197984\\t198020\\t0\\t1\\t+\n",
       "8  chr1\\t223687\\t223723\\t0\\t1\\t-\n",
       "9  chr1\\t234721\\t234757\\t0\\t1\\t+"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1\\t67926\\t67962\\t0\\t1\\t+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1\\t97370\\t97406\\t0\\t1\\t+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1\\t101069\\t101105\\t0\\t1\\t-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1\\t132314\\t132350\\t0\\t1\\t-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1\\t156509\\t156545\\t0\\t1\\t-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1\\t157123\\t157159\\t0\\t1\\t+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1\\t165094\\t165130\\t0\\t1\\t+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1\\t197984\\t198020\\t0\\t1\\t+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1\\t223687\\t223723\\t0\\t1\\t-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1\\t234721\\t234757\\t0\\t1\\t+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "metadata": {},
     "execution_count": 5
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "source": [
    "# sequence file\r\n",
    "\r\n",
    "path_fna = \"./Rattus norvegicus genome/GCA_015227675.2_mRatBN7.2_genomic.fna\"\r\n",
    "sequence = \"\"\r\n",
    "seq = \"\"\r\n",
    "with open(path_fna) as f:\r\n",
    "    lines = f.readlines()\r\n",
    "    # cut off the first line of the sequence file\r\n",
    "    lines = lines[1:]\r\n",
    "for line in lines:\r\n",
    "    sequence += line[:-1]\r\n",
    "f.close()\r\n",
    "# print(sequence)\r\n",
    "\r\n",
    "# seq with re\r\n",
    "#     sequence = f.read()\r\n",
    "# # print(sequence)\r\n",
    "# match = re.search(r\"(\\w+)(\\n)(\\w+)\", sequence)\r\n",
    "# if match:\r\n",
    "#     seq = match.groups(1)\r\n",
    "# print(seq)\r\n",
    "\r\n",
    "# seq as a df\r\n",
    "# df_seq = pd.read_csv(path_fna, header=None, sep='\\n')\r\n",
    "# pd.set_option('display.max_colwidth', None)\r\n",
    "# df_seq.head(10)\r\n"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "source": [
    "'''\r\n",
    "UTILITY VARS AND FUNCTIONS\r\n",
    "'''\r\n",
    "\r\n",
    "bases = ['C', 'G', 'T', 'A']\r\n",
    "\r\n",
    "# 1. TAKE COMPLEMENT OF A SEQUENCE\r\n",
    "def take_complement(seq=\"\"):\r\n",
    "    new_seq = \"\"\r\n",
    "    for char in seq:\r\n",
    "        if char == 'A' or char == 'a':\r\n",
    "            new_seq += 'T'\r\n",
    "        elif char == 'T' or char == 't':\r\n",
    "            new_seq += 'A'\r\n",
    "        elif char == 'G' or char == 'g':\r\n",
    "            new_seq += 'C'\r\n",
    "        elif char == 'C' or char == 'c':\r\n",
    "            new_seq += 'G'\r\n",
    "    return new_seq\r\n",
    "\r\n",
    "# 2. REVERSE STRING\r\n",
    "def reverse_str(str):\r\n",
    "    # print(\"str=\",str,\" reversed=\",str[::-1])\r\n",
    "    return str[::-1]\r\n",
    "# def reverse_str(str):\r\n",
    "#     new_str = \"\"\r\n",
    "#     size = len(str)\r\n",
    "#     for idx in range(size):\r\n",
    "#         new_str += str[size-idx-1:size-idx]\r\n",
    "#     # print(\"str=\",str,\" reversed=\",new_str)\r\n",
    "#     return new_str\r\n",
    "\r\n",
    "# 3. TO CAPS\r\n",
    "def to_caps(seq_arr):\r\n",
    "    new_seq_arr = list()\r\n",
    "    for i in range(len(seq_arr)):\r\n",
    "        new_seq_arr.append(\"\")\r\n",
    "        for j in range(len(seq_arr[i])):\r\n",
    "            if seq_arr[i][j] == \"a\":\r\n",
    "                new_seq_arr[i] += \"A\"\r\n",
    "            elif seq_arr[i][j] == \"t\":\r\n",
    "                new_seq_arr[i] += \"T\"\r\n",
    "            elif seq_arr[i][j] == \"c\":\r\n",
    "                new_seq_arr[i] += \"C\"\r\n",
    "            elif seq_arr[i][j] == \"g\":\r\n",
    "                new_seq_arr[i] += \"G\"\r\n",
    "            else:\r\n",
    "                new_seq_arr += seq_arr[i][j]\r\n",
    "            # print(j, end=\" \")\r\n",
    "    return new_seq_arr"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "PARSING DATA"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "source": [
    "'''\r\n",
    "PARSE BED FILES\r\n",
    "'''\r\n",
    "\r\n",
    "threshold = 18                   # each line of seq_arr is approximately 37 chars\r\n",
    "\r\n",
    "df_bed = df_bed_chr1\r\n",
    "seq_arr_chr1 = list()\r\n",
    "seq_arr = seq_arr_chr1\r\n",
    "\r\n",
    "labels = list()\r\n",
    "entries = list()                  # list( dict( str: list(str, str, str) ) )\r\n",
    "\r\n",
    "# Example: line = [\"chr1\", \"67926\", \"67962\", \"0\", \"1\", \"+\"]\r\n",
    "for line in df_bed.to_numpy():\r\n",
    "    # if line and len(line) > 1:\r\n",
    "    line = line.tolist()[0].split(\"\\t\")\r\n",
    "    # print(\"line=\", line)\r\n",
    "    if line[0] not in labels:\r\n",
    "        labels.append(line[0])\r\n",
    "    d = dict()\r\n",
    "    #------------------------------------------------------------\r\n",
    "    # dict = {chr#, [begin, end, +/-]}\r\n",
    "    #------------------------------------------------------------\r\n",
    "    d[line[0]] = {line[1], line[2], line[5]}\r\n",
    "    entries.append(d)\r\n",
    "\r\n",
    "    if line[5] == \"+\":\r\n",
    "        sub_seq = sequence[int(line[1]) : int(line[2])+1]\r\n",
    "        if len(sub_seq) >= threshold:\r\n",
    "            seq_arr.append(sub_seq)\r\n",
    "        # print(seq_arr)\r\n",
    "    elif line[5] == \"-\":\r\n",
    "        sub_seq = sequence[int(line[1]) : int(line[2])+1]\r\n",
    "        # sub_seq.reverse()\r\n",
    "        if len(sub_seq) >= threshold:\r\n",
    "            reversed = reverse_str(sub_seq)\r\n",
    "            reversed_complement = take_complement(reversed)\r\n",
    "            seq_arr.append(reversed_complement)\r\n",
    "    else:\r\n",
    "        seq_arr.append(\"-\")\r\n",
    "    # seq_arr = to_caps(seq_arr)\r\n",
    "\r\n",
    "print(seq_arr[-20:])\r\n"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "['AAAGTCTGGGGGTCTGCTAAGGTCTAGGTCCTGTTAG', 'AATGAATTAAGTGACAATAATATGAACTTCAACAGCA', 'TCTCCAGGCCTGTTAAAGGTTGAAATGAATTAAGTGA', 'TAGCACTTACTAAGTCTATATTTCACCTTTCCTGTCC', 'TGTCAAGGGAAAAGGGCTCCATGGGAGGGCTGCCCAG', 'CATGCATGGGAACATGTGTACAGATGCCTCTTGCTAC', 'CAATTCTACTACATTTTCTCTCAATAGCCAAATGCAT', 'CTAATCTGGGCACTGGATAAATAGGATCCTCTCTAGT', 'CATCTCTCTATGTTCTCCATGGGAGATTTCAATAACT', 'GTTACCCGACAAATATGACCAAGAAAAAGAGTGAGGA', 'GTAGATCCTCAAGATGTGAGATGCTGGAATGCCACAG', 'AGGCAGTAAGTGGATCCTAAGCAAGCTAAACATGGAC', 'AGGAACATCTGGGGCTGGCAAGTTGCTGGAGCATTTT', 'AATTGTGAGTTTTGGTCTCCTTGTCTCCCCAAGTCAC', 'CCTCCTAAGTTACACTGTGGCTAGCCCTGGGTTTTCA', 'attgtgaacatctctacatccaggtgaggcaatactt', 'GCTGCCCAGCCACAAAGTATTGCCTCACCTGGATGTA', 'GATGAAACAGGCAGTTTTGTCCAGTGGCTGCCCAGCC', 'tgatgtcagaggtgaggcaggtacaagatagaaggag', 'ccagttttctgaggaaacaccagattgacttccagag']\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "DATA CLEANING"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "# new_seq_arr = list()\r\n",
    "\r\n",
    "# # TO CAPS\r\n",
    "# for i in range(len(seq_arr)):\r\n",
    "#     new_seq_arr.append(\"\")\r\n",
    "#     for j in range(len(seq_arr[i])):\r\n",
    "#         # print(char)\r\n",
    "#         if seq_arr[i][j] == \"a\":\r\n",
    "#             new_seq_arr[i] += \"A\"\r\n",
    "#         elif seq_arr[i][j] == \"t\":\r\n",
    "#             new_seq_arr[i] += \"T\"\r\n",
    "#         elif seq_arr[i][j] == \"c\":\r\n",
    "#             new_seq_arr[i] += \"C\"\r\n",
    "#         elif seq_arr[i][j] == \"g\":\r\n",
    "#             new_seq_arr[i] += \"G\"\r\n",
    "#         else:\r\n",
    "#             new_seq_arr += seq_arr[i][j]\r\n",
    "\r\n",
    "# DELETE INVALID ENTRIES\r\n",
    "\r\n",
    "\r\n",
    "\r\n",
    "\r\n",
    "# seq_arr = new_seq_arr\r\n",
    "# print(seq_arr[:20])"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "SEQUENCE PROFILING"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "source": [
    "'''\r\n",
    "SEQUENCE PROFILING\r\n",
    "'''\r\n",
    "\r\n",
    "mat = np.asarray(seq_arr)\r\n",
    "profile = list()\r\n",
    "concensus = \"\"\r\n",
    "len_motif = len(seq_arr[0])  # aka. y in shape(x, y)\r\n",
    "\r\n",
    "for i in range(len_motif):\r\n",
    "    # nucleotides = ''.join(x[i] for x in mat)\r\n",
    "    nucleotides = list()\r\n",
    "    for alphabet in seq_arr:\r\n",
    "        if not alphabet or i >= len(alphabet):\r\n",
    "            break\r\n",
    "        nucleotides.append(alphabet[i])\r\n",
    "    count = [nucleotides.count('A'), nucleotides.count('G'), \\\r\n",
    "        nucleotides.count('C'), nucleotides.count('T')]\r\n",
    "    max_count_base = np.argmax(count)\r\n",
    "    # Lapalacian correction\r\n",
    "    for c in count:\r\n",
    "        c += 1\r\n",
    "    profile.append(count / np.sum(count))\r\n",
    "    concensus += bases[max_count_base]\r\n"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "# '''\r\n",
    "# PARSE FNA FILE\r\n",
    "# CONVERT THE DATA FRAME TO A NUMPY ARRAY:\r\n",
    "\r\n",
    "# array([['@SRR13062383.1 1 length=150'], ['CCCTGCCATCCGGAACCACGACAAGGTCCTGCAGCGCCTGTGTGAGCACTTGGACCACGCCCTGTTGTACGGATTGCAAGACCTGTCCTCTGGCTACTGGGTGGTCGTGGTCCATTTCACCCGCAGAGAGGCCATCAGGCAGATCGGAAG'],\r\n",
    "\r\n",
    "# ['+SRR13062383.1 1 length=150'],['BBBDFFFFDHHFDCGHGIJGIIJJFDFHHIIJJHHGIJHIGIHIJJJIJGGHHHHFEFDDDDBDDDDCCDDBDDDDDDDCDDDDDDDDDDDDDDDDDDCDD95>5CD@DB?CCCEEEEDDDDDDDDDDCDBDDDDDDCCDDDDDDDDDDB'],\r\n",
    "\r\n",
    "# ...\r\n",
    "\r\n",
    "# ])\r\n",
    "\r\n",
    "# '''\r\n",
    "\r\n",
    "# SEPARATE HEADER LINES AND SEQUENCE LINES\r\n",
    "\r\n",
    "# headers = list()\r\n",
    "# sequence = list()\r\n",
    "\r\n",
    "# for i in range(df.to_numpy().shape[0]):\r\n",
    "#     # print(type(df.to_numpy()[i].tolist()[0]))\r\n",
    "#     if i % 2 == 0:\r\n",
    "#         headers.append(df.to_numpy()[i].tolist()[0])\r\n",
    "#     else:\r\n",
    "#         sequence.append(str(df.to_numpy()[i].tolist()[0]))\r\n",
    "# headers = np.asarray(headers)\r\n",
    "# sequence = np.asarray(sequence)\r\n",
    "\r\n",
    "# headers\r\n",
    "# sequence\r\n",
    "\r\n",
    "\r\n",
    "\r\n",
    "\r\n",
    "\r\n",
    "# # CHECK SEQUENCE VALIDITY (DELETE NON-ALPHABETIC SEQUENCES)\r\n",
    "\r\n",
    "# valid_chars = [\"A\", \"T\", \"C\", \"G\"]\r\n",
    "\r\n",
    "# # eliminate non-alphabets\r\n",
    "# # for idx in range(len(sequence)):\r\n",
    "# idx = 0\r\n",
    "# while idx <= len(sequence):\r\n",
    "#     if idx >= len(sequence):\r\n",
    "#         break\r\n",
    "#     for char in sequence[idx]:\r\n",
    "#         match = re.match('[A-z]', char)\r\n",
    "#         # if char not in alphabets:\r\n",
    "#         if not match:\r\n",
    "#             headers = np.delete(headers, idx)\r\n",
    "#             sequence = np.delete(sequence, idx)\r\n",
    "#             idx -= 1\r\n",
    "#             break\r\n",
    "#     idx += 1\r\n",
    "\r\n",
    "# sequence"
   ],
   "outputs": [],
   "metadata": {}
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "2405ebe63de0cd10beb37bc826a8f97ace1f77a25f7bb4a1a60b2732a6e1101d"
  },
  "kernelspec": {
   "display_name": "Python 3.7.3 32-bit",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "orig_nbformat": 2
 },
 "nbformat": 4,
 "nbformat_minor": 2
}